Mathlib/LinearAlgebra/Matrix/HermitianFunctionalCalculus.lean

```diff
@@ -0,0 +1,302 @@
+  constructor
+  · contrapose!
+    intro hμ h_eig
+    have h_iSup : ⨆ μ ∈ Set.range d, eigenspace (toLin' (diagonal d)) μ = ⊤ := by
+      rw [eq_top_iff, ← (Pi.basisFun R n).span_eq, Submodule.span_le]
+      rintro - ⟨i, rfl⟩
+      simp only [SetLike.mem_coe]
+      apply Submodule.mem_iSup_of_mem (d i)
+      apply Submodule.mem_iSup_of_mem ⟨i, rfl⟩
+      rw [mem_eigenspace_iff]
+      exact (hasEigenvector_toLin'_diagonal d i).apply_eq_smul
+    have hμ_not_mem : μ ∉ Set.range d := by simpa using fun i ↦ (hμ i)
+    have := eigenspaces_independent (toLin' (diagonal d)) |>.disjoint_biSup hμ_not_mem
+    rw [h_iSup, disjoint_top] at this
+    exact h_eig this
+  · rintro ⟨i, rfl⟩
+    exact this i
+
+/-- The spectrum of the diagonal operator is the range of the diagonal viewed as a function. -/
+lemma Matrix.spectrum_diagonal (d : n → R) :
+    spectrum R (diagonal d) = Set.range d := by
+  ext μ
+  rw [← AlgEquiv.spectrum_eq (Matrix.toLinAlgEquiv <| Pi.basisFun R n),
+    ← hasEigenvalue_iff_mem_spectrum, Set.mem_range]
+  exact Matrix.hasEigenvalue_toLin'_diagonal_iff d
+
+end SpectrumDiagonal
+namespace Matrix
+
+variable {𝕜 : Type*} [RCLike 𝕜] {n : Type*} [Fintype n]
+
+namespace IsHermitian
+
+variable [DecidableEq n]
+
+variable {A : Matrix n n 𝕜} (hA : IsHermitian A)
+
+/--Eigenvalues of a Hermitian Matrix, coerced, belong to the spectrum of the assoc.toEuclideanLin -/
+theorem eigenvalue_mem_toEuclideanLin_spectrum_RCLike (i : n) :
+    (RCLike.ofReal ∘ hA.eigenvalues) i ∈ spectrum 𝕜 (toEuclideanLin A) :=
+  LinearMap.IsSymmetric.hasEigenvalue_eigenvalues _ _ _ |>.mem_spectrum
+
+/-- Algebra equivalence between the linear maps and continuous linear maps on a finite-dim module.-/
+def AlgEquivFiniteDimNormedLinearCLM.{v} (E : Type v) [NormedAddCommGroup E]
+    [NormedSpace 𝕜 E][FiniteDimensional 𝕜 E] :
+    AlgEquiv (R := 𝕜) (A := E →ₗ[𝕜] E) (B := E →L[𝕜] E) :=
+    {LinearMap.toContinuousLinearMap with
+    map_mul' := fun _ _ ↦ rfl
+    commutes' := fun _ ↦ rfl}
+
+/--Spectrum of a Hermitian matrix equals the spectrum as a EuclideanLin. -/
+theorem spec_toEuclideanLin_eq_spec : spectrum 𝕜 (toEuclideanLin A) = spectrum 𝕜 A :=
+  AlgEquiv.spectrum_eq ((AlgEquiv.trans ((toEuclideanCLM : Matrix n n 𝕜 ≃⋆ₐ[𝕜]
+  EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n) : Matrix n n 𝕜 ≃ₐ[𝕜]
+  EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n))
+  (AlgEquivFiniteDimNormedLinearCLM (EuclideanSpace 𝕜 n)).symm) _
+
+/--Eigenvalues of a hermitian matrix A are in the ℝ spectrum of A. -/
+theorem eigenvalue_mem_real : ∀ (i : n), (hA.eigenvalues) i ∈ spectrum ℝ A := by
+  intro i
+  apply spectrum.of_algebraMap_mem (S := 𝕜) (R := ℝ) (A := Matrix n n 𝕜)
+  rw [← spec_toEuclideanLin_eq_spec]
+  apply hA.eigenvalue_mem_toEuclideanLin_spectrum_RCLike i
+
+/--Definition of the StarAlgHom for the continuous functional calculus of a Hermitian matrix. -/
+@[simps]
+noncomputable def φ : StarAlgHom ℝ C(spectrum ℝ A, ℝ) (Matrix n n 𝕜) where
```

Comment:
This shouldn't be called `phi` it either needs to be inclined, or given a more informative name. Also, have you considered defining it on bare functions instead of continuous ones?

Reply:
I doubt it will clean up anything, but the advantage is that in the future we can write a lemma of the form `cfc f a = φ f` for *any* `f`, so long as `a` is selfadjoint. This is nice because the first argument to `cfc` is a bare function.

Reply:
What do you mean by 'inclined' in this context? (I am inclined to give this a more informative name...)

Reply:
I've changed the name to `cfc`. Now I have to look into changing this to bare functions...

Reply:
It seems that if I replace the type of continuous functions to the type of bare functions, the instance declaration has several places where it expects a map on continuous functions.